### PR TITLE
Support linux-cip 5.10

### PR DESCRIPTION
**Pull request by iwanami-lin**
Add 5.10 repository.
  conf/distro/emlinux.conf
  recipes-kernel/linux/linux-cip_5.10.bb
  recipes-kernel/linux-libc-headers/
Fix compile error.
  recipes-debian/qemu/qemu/
Fix SDK installer.
  recipes-images/images/core-image-sdk.inc
  recipes-kernel/linux/kernel-devsrc.bbappend

# Purpose of pull request

prototype for linux-cip 5.10 support.

# Test
## How to test
```
'# bitbake core-image-minimal
'# bitbake core-image-minimal-sdk -c populate_sdk
and install sdk.
```

## Test result
complete compile, and sdk install was succeeded.

### CODE REVIEW

File: recipes-debian/qemu/qemu_debian.bbappend [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FILESEXTRAPATHS_prepend := "${THISDIR}/qemu:"
+#FILESPATH_append += " ${THISDIR}/qemu"
```
Comment (masami256) [inline recipes-debian/qemu/qemu_debian.bbappend]: Could you remove this line?

File: recipes-images/images/core-image-sdk.inc [hunk 2]
```diff
@@ -7,7 +7,12 @@ sdk_post_process () {
...
 
         # Set up kernel headers for building user application
-        $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make mrproper; make headers_check; make INSTALL_HDR_PATH=${OECORE_TARGET_SYSROOT}/usr headers_install' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
+	if [ "${LINUX_GIT_BRANCH}" = "linux-5.10.y-cip" ]; then
+	    $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && cp include/config/kernel.release _kernel.release; cp include/generated/utsrelease.h _utsrelease.h; make prepare; mv _kernel.release include/config/kernel.release; mv _utsrelease.h include/generated/utsrelease.h; make headers_check; make INSTALL_HDR_PATH=${OECORE_TARGET_SYSROOT}/usr headers_install' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
+	else
+            $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make mrproper; make headers_check; make INSTALL_HDR_PATH=${OECORE_TARGET_SYSROOT}/usr headers_install' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
```
Comment (masami256) [inline recipes-images/images/core-image-sdk.inc]: indent is mismatch.

File: recipes-kernel/linux/kernel-devsrc.bbappend [hunk 3]
```diff
@@ -0,0 +1,216 @@
+do_install() {
```
Comment (masami256) [inline recipes-kernel/linux/kernel-devsrc.bbappend]: cp command may cause of ``` host-user-contaminated ``` QA issue[1], so using ''' cp -r --no-preserve=ownership ''' or install command is preferred. 

1: https://github.com/meta-debian/meta-debian/pull/280

File: recipes-kernel/linux/linux-cip_5.10.bb [hunk 4]
```diff
@@ -0,0 +1,96 @@
+require recipes-kernel/linux/linux-base_git.bb
```
Comment (masami256) [inline recipes-kernel/linux/linux-cip_5.10.bb:1]: Does 5.10 kernel need to support genericx86-64 too?

File: recipes-images/images/core-image-sdk.inc [hunk 5]
```diff
@@ -7,7 +7,12 @@ sdk_post_process () {
         $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make scripts' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
 
         # Set up kernel headers for building user application
-        $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make mrproper; make headers_check; make INSTALL_HDR_PATH=${OECORE_TARGET_SYSROOT}/usr headers_install' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
+	if [ "${LINUX_GIT_BRANCH}" = "linux-5.10.y-cip" ]; then
+            $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && cp include/config/kernel.release _kernel.release; cp include/generated/utsrelease.h _utsrelease.h; make prepare; mv _kernel.release include/config/kernel.release; mv _utsrelease.h include/generated/utsrelease.h; make headers_check; make INSTALL_HDR_PATH=${OECORE_TARGET_SYSROOT}/usr headers_install' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
```
Comment (Yoshihiro-Okada-ML) [inline recipes-images/images/core-image-sdk.inc]: Use this patch for both kernel-5.10.y and kernel-4.19.y because we shouldn't use make mrproper.

File: conf/distro/emlinux-cip.conf [hunk 6]
```diff
@@ -0,0 +1,10 @@
+include conf/distro/include/linux.inc
```
Comment (hiraku-wfs) [inline conf/distro/emlinux-cip.conf]: At first, please rename the emlinux-cip.conf to emlinux-k510.conf.

In addtion, please rename linux.inc to emlinux.inc.
I think you don't need the path "conf/distro/" because both files are located on the same layer.

File: conf/distro/emlinux-cip.conf [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+DISTRO = "emlinux-cip"
+
+LINUX_GIT_BRANCH ?= "linux-5.10.y-cip"
+LINUX_GIT_SRCREV ?= "3ddbe9bf6a006b50d35887b00f96c5768a32b7f3"
+LINUX_CVE_VERSION ??= "5.10.8"
+
+# preferred providers and versions
+require conf/distro/include/emlinux-preferred-provider-cip.inc
```
Comment (hiraku-wfs) [inline conf/distro/emlinux-cip.conf]: please rename emlinux-preferred-provider-cip.inc to emlinux-k510-preferred-provider.inc.

File: conf/distro/emlinux-cip.conf [hunk 8]
```diff
@@ -0,0 +1,10 @@
+include conf/distro/include/linux.inc
+
+DISTRO = "emlinux-cip"
```
Comment (hiraku-wfs) [inline conf/distro/emlinux-cip.conf]: emlinux-k510

File: conf/distro/emlinux.conf [hunk 9]
```diff
@@ -1,91 +1,16 @@
...
 #
 # If you want to use latest revision of the kernel, append the following line
 # to local.conf
 #
-# LINUX_GIT_SRCREV = "${LINUX_GIT_BRANCH}"
-#
-LINUX_GIT_SRCREV ?= "e48c182113555669a03ec13050eed01d1dc66e9f"
-LINUX_CVE_VERSION ?= "${@bb.utils.contains('LINUX_GIT_SRCREV', 'e48c182113555669a03ec13050eed01d1dc66e9f', '4.19.177', '${PV}', d)}"
-LINUX_CIP_VERSION ?= "${@bb.utils.contains('LINUX_GIT_SRCREV', 'e48c182113555669a03ec13050eed01d1dc66e9f', 'v4.19.177-cip44', '${PV}', d)}"
+LINUX_GIT_SRCREV = "${LINUX_GIT_BRANCH}"
```
Comment (hiraku-wfs) [inline conf/distro/emlinux.conf]: Why is this necessary?

File: conf/distro/emlinux.conf [hunk 10]
```diff
@@ -1,91 +1,16 @@
...
-ftp://.*/.*      http://downloads.yoctoproject.org/mirror/sources/ \n \
-http://.*/.*     http://downloads.yoctoproject.org/mirror/sources/ \n \
-https://.*/.*    http://downloads.yoctoproject.org/mirror/sources/ \n"
-
-SANITY_TESTED_DISTROS ?= " \
-            ubuntu-18.04 \n \
-            "
+LINUX_4.19_GIT_SRCREV ?= "e48c182113555669a03ec13050eed01d1dc66e9f"
+LINUX_GIT_SRCREV ?= "${LINUX_4.19_GIT_SRCREV}"
+LINUX_CVE_VERSION ??= "${@bb.utils.contains('LINUX_GIT_SRCREV', '${LINUX_4.19_GIT_SRCREV}', '4.19.177', '${PV}', d)}"
```
Comment (hiraku-wfs) [inline conf/distro/emlinux.conf]: Why didn't you make these the same as 5.10 distro?

File: recipes-debian/qemu/qemu/0014-linux-user-fix-to-handle-variably-sized-SIOCGSTAMP-w-custom.patch [hunk 11]
```diff
@@ -0,0 +1,342 @@
+This patch was copied from poky in thud branch.
```
Comment (hiraku-wfs) [inline recipes-debian/qemu/qemu/0014-linux-user-fix-to-handle-variably-sized-SIOCGSTAMP-w-custom.patch]: Please prepend # for this comment and add a blank line after that.

File: recipes-images/images/core-image-sdk.inc [hunk 12]
```diff
@@ -1,13 +1,23 @@
...
+# Post process after installed sdk
+#sdk_post_process () {
+#        # Set up kernel for building kernel module now
+#        echo "configuring scripts of kernel source for building .ko file..."
+#        $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make scripts' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
+
+#        # Set up kernel headers for building user application
+#        $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && cp include/config/kernel.release _kernel.release; cp include/generated/utsrelease.h _utsrelease.h; make prepare; mv _kernel.release include/config/kernel.release; mv _utsrelease.h include/generated/utsrelease.h; make headers_check; make INSTALL_HDR_PATH=${OECORE_TARGET_SYSROOT}/usr headers_install' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
+#}
+
```
Comment (hiraku-wfs) [inline recipes-images/images/core-image-sdk.inc]: Please remove the dead code. You can see it again by "git-log -p", etc.

File: recipes-images/images/core-image-sdk.inc [hunk 13]
```diff
@@ -1,13 +1,23 @@
...
 # Post process after installed sdk
 sdk_post_process () {
         # Set up kernel for building kernel module now
-        echo "configuring scripts of kernel source for building .ko file..."
-        $SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make scripts' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
+        #$SUDO_EXEC bash -c 'source "$0" && cd "${OECORE_TARGET_SYSROOT}/usr/src/kernel" && make scripts' $target_sdk_dir/environment-setup-@REAL_MULTIMACH_TARGET_SYS@
```
Comment (hiraku-wfs) [inline recipes-images/images/core-image-sdk.inc]: Same as above.

File: recipes-kernel/linux/kernel-devsrc.bbappend [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        if [ "${ARCH}" = "arm64" ]; then
+            cp -a --parents arch/arm64/kernel/vdso/*gettimeofday.* $kerneldir/build/
+        fi
+        cp -a --parents lib/vdso/gettimeofday.* $kerneldir/build/
+
+        cd ${B}
+        cp -a --parents scripts/module.lds $kerneldir/build/
+    else
+        cd ${S}
+        pwd
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/kernel-devsrc.bbappend]: Is this "pwd" necessary?

File: recipes-kernel/linux/linux-cip_5.10.bb [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+SRC_URI_append_raspberrypi3-64 += "file://raspberrypi3-64.config"
+LINUX_DEFCONFIG_raspberrypi3-64 = "defconfig"
+KERNEL_IMAGETYPE_raspberrypi3-64 = "Image"
+KERNEL_DEVICETREE_raspberrypi3-64 = "broadcom/bcm2837-rpi-3-b-plus.dtb \
+                                     broadcom/bcm2837-rpi-3-b.dtb"
+
+SRC_URI_append_beaglebone += "file://beaglebone.config"
+
+CVE_VERSION = "${LINUX_CVE_VERSION}"
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/linux-cip_5.10.bb]: Please put the line 6-9,15-26 into the same place between linux-base_git.bbappend and linux-cip_5.10.bb.
Create new linux-common.inc and require it from the both files.

File: recipes-kernel/linux/linux-cip_5.10.bb [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+PROVIDES += " linux-cip"
+LIC_FILES_CHKSUM = "file://COPYING;md5=6bc538ed5bd9a7fc9398086aedcd7e46"
+
+FILESEXTRAPATHS_prepend := "${THISDIR}/files:"
+SRC_URI_append = " \
+	file://base.config \
+"
+
+FILESEXTRAPATHS_append := ":${THISDIR}/../../../meta-debian/recipes-kernel/linux/files/"
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/linux-cip_5.10.bb]: Is the LAYERDIR_DEBIAN_debian available for this?

File: recipes-kernel/linux/linux-cip_5.10.bb [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+LINUX_DEFCONFIG_raspberrypi3-64 = "defconfig"
+KERNEL_IMAGETYPE_raspberrypi3-64 = "Image"
+KERNEL_DEVICETREE_raspberrypi3-64 = "broadcom/bcm2837-rpi-3-b-plus.dtb \
+                                     broadcom/bcm2837-rpi-3-b.dtb"
+
+SRC_URI_append_beaglebone += "file://beaglebone.config"
+
+CVE_VERSION = "${LINUX_CVE_VERSION}"
+
+#addtask shared_workdir_modules after do_compile_kernelmodules
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/linux-cip_5.10.bb]: Please add comments in detail.

File: recipes-kernel/linux/linux-cip_5.10.bb [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+do_compile_kernelmodules_append () {
+#do_shared_workdir_modules () {
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/linux-cip_5.10.bb]: Please remove unused code.

File: recipes-kernel/linux/kernel-devsrc.bbappend [hunk 19]
```diff
@@ -0,0 +1,27 @@
+do_install_prepend() {
+    cd ${S}
+    if [ ${LINUX_GIT_BRANCH} = "linux-5.10.y-cip" ]; then
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/kernel-devsrc.bbappend]: It's not good to rely on a particular branch name.
Can we write it like this?
`if ${@bb.utils.vercmp_string_op(d.getVar("KERNEL_VERSION"), '5.10', '>=')}; then`

I think that a valid version will be available for the KERNEL_VERSION because it is used after do_configure().
https://www.yoctoproject.org/docs/2.3/ref-manual/ref-manual.html#var-KERNEL_VERSION

File: conf/distro/emlinux-cip.conf (hunk 6 above)
Comment (hiraku-wfs) [reply conf/distro/emlinux-cip.conf]: and please use 'require' instead of  'include'.

File: conf/distro/emlinux.conf [hunk 20]
```diff
@@ -1,91 +1,16 @@
...
-LINUX_GIT_PREFIX ?= ""
-LINUX_GIT_REPO ?= "linux-cip.git"
+DISTRO = "emlinux"
 LINUX_GIT_BRANCH ?= "linux-4.19.y-cip"
 #
 # If you want to use latest revision of the kernel, append the following line
 # to local.conf
 #
-# LINUX_GIT_SRCREV = "${LINUX_GIT_BRANCH}"
-#
```
Comment (hiraku-wfs) [inline conf/distro/emlinux.conf:38]: Please keep this comment lines.

File: conf/distro/emlinux.conf [hunk 21]
```diff
@@ -1,91 +1,14 @@
...
-
-MIRRORS =+ "\
-ftp://.*/.*      http://downloads.yoctoproject.org/mirror/sources/ \n \
-http://.*/.*     http://downloads.yoctoproject.org/mirror/sources/ \n \
-https://.*/.*    http://downloads.yoctoproject.org/mirror/sources/ \n"
-
-SANITY_TESTED_DISTROS ?= " \
-            ubuntu-18.04 \n \
-            "
+#LINUX_GIT_SRCREV = "${LINUX_GIT_BRANCH}"
```
Comment (hiraku-wfs) [inline conf/distro/emlinux.conf:11]: I think you forgot to remove this.

File: recipes-kernel/linux/kernel-devsrc.bbappend [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+do_install_append() {
+    if [ "${@bb.utils.vercmp_string_op(d.getVar("KERNEL_VERSION"), '5.10', '>=')}" = "True" ]; then
+        cd ${S}
+        if [ "${ARCH}" = "arm64" ]; then
+            install -D arch/arm64/kernel/vdso/*gettimeofday.* \
+                $kerneldir/build/arch/arm64/kernel/vdso/
+        fi
+        install -D lib/vdso/gettimeofday.* $kerneldir/build/lib/vdso/
+        install -D K* $kerneldir/build/
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/kernel-devsrc.bbappend]: Please comment that you intend Kbuild and Kconfig (or explicitly specify these names).

File: recipes-kernel/linux/linux-k510.bb [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+CVE_VERSION = "${LINUX_CVE_VERSION}"
+
+#fix me
+#Correctry, do_shared_workdir_modules should be created and executed after 
+#do_compile_kernelmodules. 
+#However, Yocto did not recognize created task. 
+#Therefore, added operations for shared_workdir to the end of do_compile_kernelmodules.
+#addtask shared_workdir_modules after do_compile_kernelmodules
+#do_shared_workdir_modules () { }
```
Comment (hiraku-wfs) [inline recipes-kernel/linux/linux-k510.bb]:
```
# FIXME: A new task should be created to copy scripts/module.lds after do_compile_kernelmodules(),
# but the poky did not recognize the new task:
#   addtask shared_workdir_modules after do_compile_kernelmodules
# So we decided to append the task to the do_compile_kernelmodules().
```

File: conf/distro/emlinux.conf (hunk 21 above)
Comment (hiraku-wfs) [reply conf/distro/emlinux.conf:11]: Sorry, it's my mis-understanding.

Comment (hiraku-wfs) [reply conf/distro/emlinux.conf:11]: Resolved

**Comment by hiraku-wfs** [comment]
#113 was merged instead of this PR.
